src/common/src/util/value_encoding/column_aware_row_encoding.rs

```diff
@@ -313,76 +314,124 @@ impl RowEncoding {
         self.header.set_offset(offset_width);
 
         self.offsets
-            .reserve_exact(usize_offsets.len() * offset_width.width());
+            .resize(usize_offsets.len() * offset_width.width(), 0);
+
+        let mut offsets_buf = &mut self.offsets[..];
         for &offset in usize_offsets {
-            self.offsets
-                .put_uint_le(offset as u64, offset_width.width());
+            offsets_buf.put_uint_le(offset as u64, offset_width.width());
         }
     }
 
-    fn encode<T: Encode>(&mut self, datums: impl IntoIterator<Item = T>, data_types: &[DataType]) {
+    fn encode<T: Encode>(
+        &mut self,
+        datums: impl IntoIterator<Item = T>,
+        data_types: impl IntoIterator<Item = &DataType>,
+    ) {
         debug_assert!(
-            self.buf.is_empty(),
+            self.data.is_empty(),
             "should not encode one RowEncoding object multiple times."
         );
         let datums = datums.into_iter();
-        let mut offset_usize = Vec::with_capacity(datums.size_hint().0);
+        let mut offset_usize =
+            SmallVec::<[usize; COLUMN_ON_STACK]>::with_capacity(datums.size_hint().0);
         for (datum, data_type) in datums.zip_eq_debug(data_types) {
-            offset_usize.push(self.buf.len());
-            datum.encode_to(data_type, &mut self.buf);
+            offset_usize.push(self.data.len());
+            datum.encode_to(data_type, &mut self.data);
         }
         self.set_offsets(&offset_usize);
     }
 }
 
+mod data_types {
+    use crate::types::{DataType, StructType};
+
+    /// A trait unifying data types of a row and field types of a struct.
+    pub trait DataTypes: Clone {
+        fn iter(&self) -> impl Iterator<Item = &DataType>;
+    }
+
+    impl DataTypes for Vec<DataType> {
+        fn iter(&self) -> impl Iterator<Item = &DataType> {
+            self.as_slice().iter()
+        }
+    }
+
+    impl DataTypes for StructType {
+        fn iter(&self) -> impl Iterator<Item = &DataType> {
+            self.types()
+        }
+    }
+}
+use data_types::DataTypes;
+
 /// Column-Aware `Serializer` holds schema related information, and shall be
 /// created again once the schema changes
 #[derive(Clone)]
-pub struct Serializer {
-    encoded_column_ids: Vec<u8>,
-    data_types: Vec<DataType>,
+pub struct Serializer<D: DataTypes = Vec<DataType>> {
+    encoded_column_ids: EncodedColumnIds,
+    data_types: D,
+}
+
+type EncodedColumnIds = SmallVec<[u8; COLUMN_ON_STACK * 4]>;
```

Comment:
So our column id is 4-byte now. Will using 2-byte id improve the performance a lot? IIRC the column id is local for every table.

Reply:
The reason for using 4 bytes is that, we are using `i32` as our internal representation of column id everywhere, and previous implementation of column-aware encoding did not adopt variable-length integer encoding as well.

Using 2 bytes (or variable) can definitely reduce the length of the encoded row, but I'm not sure it this can improves encoding procedure much. 🤔